### PR TITLE
Implement faster memcmp for x86_64

### PR DESCRIPTION
**Pull request by Demindiro**
On x86_64 it is generally fast to read unaligned words, so there is no need to do byte-by-byte comparisons.

The new algorithm compares 32 bytes at a time. If any doesn't match it effectively does a binary search to find the first non-matching byte.

Benchmark results on a Ryzen 7 2700X:

<details>
<summary><code>master</code> (with extra benchmarks)</summary>
<pre>
test memcmp_rust_1048576              ... bench:     293,975 ns/iter (+/- 3,188) = 3566 MB/s
test memcmp_rust_16                   ... bench:           7 ns/iter (+/- 0) = 2285 MB/s
test memcmp_rust_32                   ... bench:          11 ns/iter (+/- 1) = 2909 MB/s
test memcmp_rust_4096                 ... bench:       1,049 ns/iter (+/- 203) = 3904 MB/s
test memcmp_rust_64                   ... bench:          32 ns/iter (+/- 4) = 2000 MB/s
test memcmp_rust_8                    ... bench:           4 ns/iter (+/- 1) = 2000 MB/s
test memcmp_rust_unaligned_1048575    ... bench:     276,654 ns/iter (+/- 78,924) = 3790 MB/s
test memcmp_rust_unaligned_15         ... bench:           8 ns/iter (+/- 0) = 2000 MB/s
test memcmp_rust_unaligned_31         ... bench:          11 ns/iter (+/- 2) = 2909 MB/s
test memcmp_rust_unaligned_4095       ... bench:       1,234 ns/iter (+/- 301) = 3319 MB/s
test memcmp_rust_unaligned_63         ... bench:          33 ns/iter (+/- 11) = 1939 MB/s
test memcmp_rust_unaligned_7          ... bench:           4 ns/iter (+/- 0) = 2000 MB/s
</pre>
</details>

<details>
<summary><code>memcmp-x86_64</code></summary>
<pre>
test memcmp_rust_1048576              ... bench:      24,682 ns/iter (+/- 240) = 42483 MB/s
test memcmp_rust_16                   ... bench:           3 ns/iter (+/- 0) = 5333 MB/s
test memcmp_rust_32                   ... bench:           4 ns/iter (+/- 0) = 8000 MB/s
test memcmp_rust_4096                 ... bench:         113 ns/iter (+/- 1) = 36247 MB/s
test memcmp_rust_64                   ... bench:           5 ns/iter (+/- 0) = 12800 MB/s
test memcmp_rust_8                    ... bench:           3 ns/iter (+/- 0) = 2666 MB/s
test memcmp_rust_unaligned_1048575    ... bench:      27,049 ns/iter (+/- 6,989) = 38765 MB/s
test memcmp_rust_unaligned_15         ... bench:           3 ns/iter (+/- 0) = 5333 MB/s
test memcmp_rust_unaligned_31         ... bench:           4 ns/iter (+/- 0) = 8000 MB/s
test memcmp_rust_unaligned_4095       ... bench:         100 ns/iter (+/- 6) = 40960 MB/s
test memcmp_rust_unaligned_63         ... bench:           5 ns/iter (+/- 0) = 12800 MB/s
test memcmp_rust_unaligned_7          ... bench:           3 ns/iter (+/- 0) = 2666 MB/s
</pre>
</details>

<details>
<summary>builtin</summary>
<pre>
test memcmp_builtin_1048576           ... bench:      23,202 ns/iter (+/- 1,283) = 45193 MB/s
test memcmp_builtin_16                ... bench:           4 ns/iter (+/- 0) = 4000 MB/s
test memcmp_builtin_32                ... bench:           3 ns/iter (+/- 1) = 10666 MB/s
test memcmp_builtin_4096              ... bench:          73 ns/iter (+/- 16) = 56109 MB/s
test memcmp_builtin_64                ... bench:           4 ns/iter (+/- 1) = 16000 MB/s
test memcmp_builtin_8                 ... bench:           4 ns/iter (+/- 1) = 2000 MB/s
test memcmp_builtin_unaligned_1048575 ... bench:      25,353 ns/iter (+/- 793) = 41359 MB/s
test memcmp_builtin_unaligned_15      ... bench:           5 ns/iter (+/- 0) = 3200 MB/s
test memcmp_builtin_unaligned_31      ... bench:           4 ns/iter (+/- 0) = 8000 MB/s
test memcmp_builtin_unaligned_4095    ... bench:          85 ns/iter (+/- 4) = 48188 MB/s
test memcmp_builtin_unaligned_63      ... bench:           4 ns/iter (+/- 0) = 16000 MB/s
test memcmp_builtin_unaligned_7       ... bench:           4 ns/iter (+/- 0) = 2000 MB/s
</pre>
</details>

Note that the results may vary a lot, e.g. `memcmp_rust_unaligned_1048575` can be as slow as `26933 MB/s` at times.

I also added some extra tests though I don't think there are enough yet. I am not sure what edge cases should be considered though.

### CODE REVIEW

File: src/mem/x86_64.rs [hunk 1]
```diff
@@ -98,3 +100,47 @@ pub unsafe fn set_bytes(dest: *mut u8, c: u8, count: usize) {
...
+    #[inline(always)]
+    unsafe fn cmp<T, U, F>(mut a: *const T, mut b: *const T, n: usize, f: F) -> i32
+    where
+        T: Clone + Copy + Eq,
+        U: Clone + Copy + Eq,
+        F: FnOnce(*const U, *const U, usize) -> i32,
+    {
+        // Just to be sure we're actually working with powers of two...
+        let _ = const { 1 - mem::size_of::<T>().count_ones() }; // <= 1
+        let _ = const { mem::size_of::<T>().count_ones() - 1 }; // >= 1
```
Comment (Amanieu) [inline src/mem/x86_64.rs:115]: This would be much clearer as `const { assert!(mem::size_of::<T>().is_power_of_two()) }`.

File: src/mem/x86_64.rs [hunk 2]
```diff
@@ -98,3 +100,47 @@ pub unsafe fn set_bytes(dest: *mut u8, c: u8, count: usize) {
...
+
+        // This should be equivalent to division with power-of-two sizes, except the former
+        // somehow still leaves a call to panic because ??
```
Comment (Amanieu) [inline src/mem/x86_64.rs:118]: This is because of the zero-check which isn't optimized away in debug mode. You can use [`unchecked_div`](https://doc.rust-lang.org/nightly/core/intrinsics/fn.unchecked_div.html) to avoid this. Or leave the code as it is, it's fine too.
